std/experimental/normint.d

```diff
@@ -0,0 +1,876 @@
+// Written in the D programming language.
+
+/**
+This module implements support for normalized integers.
+
+Authors:    Manu Evans
+Copyright:  Copyright (c) 2015, Manu Evans.
+License:    $(WEB boost.org/LICENSE_1_0.txt, Boost License 1.0)
+Source:     $(PHOBOSSRC std/experimental/_normint.d)
+*/
+module std.experimental.normint;
+
+import std.traits : isIntegral, isSigned, isUnsigned, isFloatingPoint, Unsigned;
+static import std.algorithm;
+
+import std.experimental.color : FloatTypeFor;
+
+@safe pure nothrow @nogc:
+
+/**
+Check if $(D_INLINECODE I) is a valid NormalizedInt type.
+Valid integers are $(D_INLINECODE (u)byte), $(D_INLINECODE (u)short), $(D_INLINECODE (u)int). $(D_INLINECODE (u)long) is not supported.
+*/
+enum isNormalizedIntegralType(I) = isIntegral!I && I.sizeof < 8;
+
+
+/**
+Normalized integers express a fractional range of values within an integer data type.
+
+Unsigned integers map the values $(D_INLINECODE [0, I.max]) to the fractional values $(D_INLINECODE [0.0, 1.0]) in equal increments.$(BR)
+Signed integers represent the values $(D_INLINECODE [-I.max, I.max]) to fractional values $(D_INLINECODE [-1.0, 1.0]) in equal increments. $(D_INLINECODE I.min) is outside the nominal integer range and is clamped to represent $(D_INLINECODE -1.0).
+
+Params: $(D_INLINECODE I) = $(D_INLINECODE (u)byte), $(D_INLINECODE (u)short), $(D_INLINECODE (u)int). $(D_INLINECODE (u)long) is not supported.
+*/
+struct NormalizedInt(I) if (isNormalizedIntegralType!I)
+{
+@safe:
+
+    string toString() const
+    {
+        import std.conv : to;
+        return to!string(cast(double)this);
+    }
+
+pure nothrow @nogc:
+
+    I value;
```

Comment:
These are particular complexities of implementing the png format though. I don't think it's right for colour to compromise it's type safety or correct-ness for a particular bit of code that will be complex regardless; png is a very complex format. Accessing the raw channel data is possible, I just don't think it should be encouraged, because unless the user knows precisely what they're doing, there is high probability they will handle it incorrectly.